wellcomeml/ml/bert_semantic_equivalence.py

```diff
@@ -0,0 +1,198 @@
+                {
+                    "input_ids": tf.TensorShape([None]),
+                    "attention_mask": tf.TensorShape([None]),
+                    "token_type_ids": tf.TensorShape([None]),
+                },
+                tf.TensorShape([]),
+            ),
+        )
+
+        return dataset
+
+    def fit(self, X, y, random_state=None, epochs=3):
+        """
+        Fits a sentence similarity model
+
+        Args:
+            X: List of 2-uples: [('Sentence 1', 'Sentence 2'), ....]
+            y: list of 0 (not similar) and 1s (similar)
+            random_state: a random state for the train_test_split
+            epochs: number_of epochs
+
+        Returns:
+
+        """
+        # Initialises/downloads model if not trained before.
+        # If trained, fits extra epochs without the transformations
+        if not self._trained:
+            self._initialise_models()
+            
+            # Train/val split
+            X_train, X_valid, y_train, y_valid = train_test_split(
+                X, y, test_size=self.test_size, random_state=random_state
+            )
+    
+            # Generates tensorflow dataset
+            self.train_dataset = self._prep_dataset_generator(X_train, y_train)
+            self.valid_dataset = self._prep_dataset_generator(X_valid, y_valid)
+    
+            # Generates mini-batches and stores in a class variable
+            self.train_dataset = self.train_dataset.shuffle(self.max_length).\
+                batch(self.batch_size).repeat(-1)
+            self.valid_dataset = self.valid_dataset.batch(self.eval_batch_size)
+    
+            opt = tf.keras.optimizers.Adam(learning_rate=self.learning_rate,
+                                           epsilon=1e-08)
+    
+            loss = tf.keras.losses.SparseCategoricalCrossentropy(
+                from_logits=True
+            )
+    
+            # Train and evaluate using tf.keras.Model.fit()
+    
+            metric = tf.keras.metrics.SparseCategoricalAccuracy("accuracy")
+    
+            self.model.compile(optimizer=opt, loss=loss, metrics=[metric])
+    
+            self.train_steps = len(X_train) // self.batch_size
+            self.valid_steps = len(X_valid) // self.eval_batch_size
+
+        history = self.model.fit(
+            self.train_dataset,
+            epochs=epochs,
+            steps_per_epoch=self.train_steps,
+            validation_data=self.valid_dataset,
+            validation_steps=self.valid_steps
+        )
+
+        # Accumulates history of metrics from different partial fits
+        for metric, values in history.history.items():
+            self.history[metric] += values
+
+        self._trained = True
+
+        return self
+
+    def score(self, X):
+        """
+        Calculates scores for model prediction
+
+        Args:
+            X: List of 2-uples: [('Sentence 1', 'Sentence 2'), ....]
+
+        Returns:
+            An array of shape len(X) x 2 with scores for classes 0 and 1
```

Comment:
doesn't this return an array of len(X)?

Reply:
No, this returns two scores for every data point (probably of 0 and probability of 1). Possibly redundant (could do with one score), but it is the usual softmax output for Keras.